### DISCUSSION TITLE
Payments source cann't be blank

### DISCUSSION THREAD

**Discussion by dominicsanto** [solidusio/solidus#4067] [Troubleshooting] [answered]
Trying to create a custom payment integration for my local payment provider but running into issues when select the new payment method:

![image](https://user-images.githubusercontent.com/22883553/118487069-a3cbb080-b71a-11eb-9cac-f89ea99ad345.png)

```
module Spree
  class Gateway::NewPayments < Spree::PaymentMethod
    def gateway_class
      Spree::Gateway::NewPayments
    end

    def provider_class
      Spree::Gateway::NewPayments
    end

    def payment_source_class
      Spree::CreditCard
    end
  end
end

```

Notes:

I need to not rely on the credit card form that shows up, I am trying to alter the flow where by when they select `Credit Card` after the confirm step it goes off to the payment provider themself and I check for a response to confirm or reject the order.

**Comment by jarednorman** [comment, marked as answer]
Your payment method should implement `source_required?` in order for Solidus to identify that (if this is the case) your payments won't be associated with some kind of source record representing the payment method.

Here's where that's called in the Payment model: https://github.com/solidusio/solidus/blob/master/core/app/models/spree/payment.rb#L185-L187

Here's the validation that's failing: https://github.com/solidusio/solidus/blob/master/core/app/models/spree/payment.rb#L41

Here's an example of a payment method that doesn't require a source: https://github.com/solidusio/solidus/blob/master/core/app/models/spree/payment_method/check.rb#L32-L34

All that said, you likely want to redirect for payment before completing the order and in so doing create a payment with a source that represents whatever makes sense for how your gateway works, but I say that without knowing anything about your specific situation, so disregard if it's not relevant.

**Reply by dominicsanto** [reply]
@jarednorman thanks for the reply. I worked off of the check payment method and then on the confirm step, after placing order I am redirecting to my payment provider.
